File: apps/envoygo/lib/services/library_read_fetch.dart
/// Phase 45C — Browser fetch helper for EnvoyGo (mirrors Social library-read-fetch).
///
/// Auto-chunks when the home returns `too_large`. Range bodies are always
/// base64 from the server. Pass [cache] + [revalidate] for If-None-Match.
library library_read_fetch;

import 'dart:convert';
import 'dart:typed_data';

import '../models/library_read.dart';

const libraryReadChunkBytes = 40 * 1024;

typedef LibraryReadFn = Future<LibraryReadResult> Function({
  required String targetOwnerId,
  required String path,
  Map<String, int>? range,
  String? ifNoneMatch,
  int? timeoutMs,
});

class BrowserFetchCacheEntry {
  final String body;
  final String contentType;
  final String contentHash;
  final String etag;
  final int byteLength;
  final bool isText;

  const BrowserFetchCacheEntry({
    required this.body,
    required this.contentType,
    required this.contentHash,
    required this.etag,
    required this.byteLength,
    required this.isText,
  });
}

class BrowserFetchResult {
  final String status; // ok | not_found | forbidden | error
  final String? body;
  final String? contentType;
  final String? contentHash;
  final String? etag;
  final int? byteLength;
  final bool isText;
  final String? error;
  final bool fromCache;

  const BrowserFetchResult({
    required this.status,
    this.body,
    this.contentType,
    this.contentHash,
    this.etag,
    this.byteLength,
    this.isText = false,
    this.error,
    this.fromCache = false,
  });
}

bool isTextMime(String mime) =>
    mime.startsWith('text/') || mime == 'application/json';

String _concatBase64Chunks(List<String> chunks) {
  final out = BytesBuilder(copy: false);
  for (final c in chunks) {
    out.add(base64Decode(c));
  }
  return base64Encode(out.takeBytes());
}

/// Fetch a library path via [libraryRead], auto-chunking on `too_large`.
Future<BrowserFetchResult> fetchLibraryContent(
  LibraryReadFn libraryRead, {
  required String targetOwnerId,
  required String path,
  BrowserFetchCacheEntry? cache,
  bool revalidate = false,
}) async {
  final ifNoneMatch =
      revalidate && cache != null && cache.etag.isNotEmpty ? cache.etag : null;

  final first = await libraryRead(
    targetOwnerId: targetOwnerId,
    path: path,
    ifNoneMatch: ifNoneMatch,
  );

  if (first.status == 'not_modified') {
    if (cache != null) {
      return BrowserFetchResult(
        status: 'ok',
        body: cache.body,
        contentType: cache.contentType,
        contentHash: cache.contentHash,
        etag: cache.etag,
        byteLength: cache.byteLength,
        isText: cache.isText,
        fromCache: true,
      );
    }
    return const BrowserFetchResult(
      status: 'error',
      error: 'not_modified without local cache',
    );
  }

  if (first.status == 'not_found' || first.status == 'forbidden') {
    return BrowserFetchResult(
      status: first.status,
      error: first.error,
    );
  }

  if (first.status == 'ok' && first.body != null && first.contentType != null) {
    return BrowserFetchResult(
      status: 'ok',
      body: first.body,
      contentType: first.contentType,
      contentHash: first.contentHash,
      etag: first.etag,
      byteLength: first.byteLength ?? first.body!.length,
      isText: isTextMime(first.contentType!),
    );
  }

  if (first.status == 'too_large') {
    final total = first.byteLength;
    if (total == null || total <= 0) {
      return const BrowserFetchResult(
        status: 'error',
        error: 'too_large without byteLength',
      );
    }
    final contentType = first.contentType ?? 'application/octet-stream';
    final text = isTextMime(contentType);
    final chunks = <String>[];
    var start = 0;
    while (start < total) {
      final end = (start + libraryReadChunkBytes - 1).clamp(0, total - 1);
      final part = await libraryRead(
        targetOwnerId: targetOwnerId,
        path: path,
        range: {'start': start, 'end': end},
      );
      if (part.status != 'ok' || part.body == null) {
        return BrowserFetchResult(
          status: 'error',
          error: part.error ??
              'range fetch failed at $start-$end (${part.status})',
        );
      }
      chunks.add(part.body!);
      start = end + 1;
    }
    final mergedB64 = _concatBase64Chunks(chunks);
    final body = text ? utf8.decode(base64Decode(mergedB64)) : mergedB64;
    return BrowserFetchResult(
      status: 'ok',
      body: body,
      contentType: contentType,
      contentHash: first.contentHash,
      etag: first.etag,
      byteLength: total,
      isText: text,
    );
  }

  return BrowserFetchResult(
    status: 'error',
    error: first.error ?? 'unexpected status ${first.status}',
  );
}

File: apps/social/src/lib/library-read-fetch.ts
/**
 * Phase 45B — Browser fetch helper for `library.read`.
 *
 * Handles:
 *   - Normal single-shot reads
 *   - Auto range assembly when the server returns `too_large`
 *   - If-None-Match revalidation (returns cached body on `not_modified`)
 */

import type { LibraryReadParams, LibraryReadResult } from "@envoymesh/api";

/**
 * Chunk size for range requests. Must stay under the server binary body cap
 * (`MAX_LIBRARY_READ_BINARY_BYTES` = 40 KiB in `library-read-inbound.ts`)
 * so base64 framing still fits the 64 KiB envelope guard.
 */
export const LIBRARY_READ_CHUNK_BYTES = 40 * 1024;

export type LibraryReadFn = (params: LibraryReadParams) => Promise<LibraryReadResult>;

export interface BrowserFetchCacheEntry {
  body: string;
  contentType: string;
  contentHash: string;
  etag: string;
  byteLength: number;
  isText: boolean;
}

export interface BrowserFetchResult {
  status: "ok" | "not_found" | "forbidden" | "not_modified" | "error";
  body?: string;
  contentType?: string;
  contentHash?: string;
  etag?: string;
  byteLength?: number;
  isText?: boolean;
  error?: string;
  /** True when the body was kept from cache after not_modified. */
  fromCache?: boolean;
}

function isTextMime(mime: string): boolean {
  return mime.startsWith("text/") || mime === "application/json";
}

function concatBase64Chunks(chunks: string[]): string {
  const parts = chunks.map((c) => {
    const bin = atob(c);
    const out = new Uint8Array(bin.length);
    for (let i = 0; i < bin.length; i++) out[i] = bin.charCodeAt(i);
    return out;
  });
  let total = 0;
  for (const p of parts) total += p.byteLength;
  const merged = new Uint8Array(total);
  let offset = 0;
  for (const p of parts) {
    merged.set(p, offset);
    offset += p.byteLength;
  }
  let s = "";
  for (let i = 0; i < merged.length; i++) s += String.fromCharCode(merged[i]!);
  return btoa(s);
}

/**
 * Fetch a library path, automatically chunking when the server reports too_large.
 * Pass `cache` + set `revalidate: true` on Reload to send If-None-Match.
 */
export async function fetchLibraryContent(
  libraryRead: LibraryReadFn,
  input: {
    targetOwnerId: string;
    path: string;
    cache?: BrowserFetchCacheEntry | null;
    revalidate?: boolean;
  },
): Promise<BrowserFetchResult> {
  const ifNoneMatch =
    input.revalidate && input.cache?.etag ? input.cache.etag : undefined;

  const first = await libraryRead({
    targetOwnerId: input.targetOwnerId,
    path: input.path,
    ifNoneMatch,
  });

  if (first.status === "not_modified" && input.cache) {
    return {
      status: "ok",
      body: input.cache.body,
      contentType: input.cache.contentType,
      contentHash: input.cache.contentHash,
      etag: input.cache.etag,
      byteLength: input.cache.byteLength,
      isText: input.cache.isText,
      fromCache: true,
    };
  }

  if (first.status === "not_found" || first.status === "forbidden") {
    return { status: first.status, error: first.error };
  }

  if (first.status === "ok" && first.body !== undefined && first.contentType) {
    return {
      status: "ok",
      body: first.body,
      contentType: first.contentType,
      contentHash: first.contentHash,
      etag: first.etag,
      byteLength: first.byteLength ?? first.body.length,
      isText: isTextMime(first.contentType),
    };
  }

  if (first.status === "too_large") {
    const total = first.byteLength;
    if (!total || total <= 0) {
      return { status: "error", error: "too_large without byteLength" };
    }
    const contentType = first.contentType ?? "application/octet-stream";
    const text = isTextMime(contentType);
    // Range responses are always base64 (server-side) to avoid mid-UTF-8 splits.
    const chunks: string[] = [];
    let start = 0;
    while (start < total) {
      const end = Math.min(start + LIBRARY_READ_CHUNK_BYTES - 1, total - 1);
      const part = await libraryRead({
        targetOwnerId: input.targetOwnerId,
        path: input.path,
        range: { start, end },
      });
      if (part.status !== "ok" || part.body === undefined) {
        return {
          status: "error",
          error: part.error ?? `range fetch failed at ${start}-${end} (${part.status})`,
        };
      }
      chunks.push(part.body);
      start = end + 1;
    }
    const mergedB64 = concatBase64Chunks(chunks);
    const body = text
      ? new TextDecoder().decode(
          Uint8Array.from(atob(mergedB64), (c) => c.charCodeAt(0)),
        )
      : mergedB64;
    return {
      status: "ok",
      body,
      contentType,
      contentHash: first.contentHash,
      etag: first.etag,
      byteLength: total,
      isText: text,
    };
  }

  return {
    status: "error",
    error: first.error ?? `unexpected status ${first.status}`,
  };
}
